<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>RNA-seq分析流程 | Eddie's Blog</title><meta name="author" content="Leihjeunyihn"><meta name="copyright" content="Leihjeunyihn"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="ffffff"><meta name="description" content="记录一下学习RNA-seq分析的过程，整体流程参考，可能会补充一些其他的资料： refer .a {    color: white;    font-size: 8px;    text-decoration: none;} refer a:hover {    color: blue;}&lt;&#x2F;style&gt;   RNA-seq入门实战（零）：RNA-seq流程前的准备——Linux与R">
<meta property="og:type" content="article">
<meta property="og:title" content="RNA-seq分析流程">
<meta property="og:url" content="https://eddieljx.github.io/2024/05/22/RNA-seq%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/index.html">
<meta property="og:site_name" content="Eddie&#39;s Blog">
<meta property="og:description" content="记录一下学习RNA-seq分析的过程，整体流程参考，可能会补充一些其他的资料： refer .a {    color: white;    font-size: 8px;    text-decoration: none;} refer a:hover {    color: blue;}&lt;&#x2F;style&gt;   RNA-seq入门实战（零）：RNA-seq流程前的准备——Linux与R">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://eddieljx.github.io/img/pagecover2.jpg">
<meta property="article:published_time" content="2024-05-22T10:57:36.000Z">
<meta property="article:modified_time" content="2024-06-10T09:20:24.000Z">
<meta property="article:author" content="Leihjeunyihn">
<meta property="article:tag" content="生信">
<meta property="article:tag" content="代码">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://eddieljx.github.io/img/pagecover2.jpg"><link rel="shortcut icon" href="/img/favicon-16x16.png"><link rel="canonical" href="https://eddieljx.github.io/2024/05/22/RNA-seq%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="stylesheet" href="/css/index.css?v=4.13.0"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.5.1/css/all.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/node-snackbar@0.1.16/dist/snackbar.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui@5.0.33/dist/fancybox/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: {"path":"/search.xml","preload":false,"top_n_per_article":1,"unescape":false,"languages":{"hits_empty":"找不到您查询的内容：${query}","hits_stats":"共找到 ${hits} 篇文章"}},
  translate: {"defaultEncoding":2,"translateDelay":0,"msgToTraditionalChinese":"繁","msgToSimplifiedChinese":"简"},
  noticeOutdate: undefined,
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":230},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: {"chs_to_cht":"你已切换为繁体中文","cht_to_chs":"你已切换为简体中文","day_to_night":"你已切换为深色模式","night_to_day":"你已切换为浅色模式","bgLight":"#49b1f5","bgDark":"#1f1f1f","position":"top-left"},
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid@4.11.1/dist/infinitegrid.min.js',
    buttonText: '加载更多'
  },
  isPhotoFigcaption: false,
  islazyload: true,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: true,
  },
  autoDarkmode: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: 'RNA-seq分析流程',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: false,
  postUpdate: '2024-06-10 17:20:24'
}</script><script>(win=>{
      win.saveToLocal = {
        set: (key, value, ttl) => {
          if (ttl === 0) return
          const now = Date.now()
          const expiry = now + ttl * 86400000
          const item = {
            value,
            expiry
          }
          localStorage.setItem(key, JSON.stringify(item))
        },
      
        get: key => {
          const itemStr = localStorage.getItem(key)
      
          if (!itemStr) {
            return undefined
          }
          const item = JSON.parse(itemStr)
          const now = Date.now()
      
          if (now > item.expiry) {
            localStorage.removeItem(key)
            return undefined
          }
          return item.value
        }
      }
    
      win.getScript = (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        script.onerror = reject
        script.onload = script.onreadystatechange = function() {
          const loadState = this.readyState
          if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
          script.onload = script.onreadystatechange = null
          resolve()
        }

        Object.keys(attr).forEach(key => {
          script.setAttribute(key, attr[key])
        })

        document.head.appendChild(script)
      })
    
      win.getCSS = (url, id = false) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onerror = reject
        link.onload = link.onreadystatechange = function() {
          const loadState = this.readyState
          if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
          link.onload = link.onreadystatechange = null
          resolve()
        }
        document.head.appendChild(link)
      })
    
      win.activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', 'ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
        if (t === 'dark') activateDarkMode()
        else if (t === 'light') activateLightMode()
      
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
      const detectApple = () => {
        if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
          document.documentElement.classList.add('apple')
        }
      }
      detectApple()
    })(window)</script><link rel="stylesheet" href="/css/iconfont.css"><link rel="stylesheet" href="/css/custom.css" media="defer" onload="this.media='all'"><!-- hexo injector head_end start -->
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.12.0/dist/katex.min.css">

<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/hexo-math@4.0.0/dist/style.css">
<link rel="stylesheet" href="/css/runtime.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-categories-card@1.0.0/lib/categorybar.css"><link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-tag-plugins-plus@latest/lib/assets/font-awesome-animation.min.css" media="defer" onload="this.media='all'"><link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-tag-plugins-plus@latest/lib/tag_plugins.css" media="defer" onload="this.media='all'"><script src="https://npm.elemecdn.com/hexo-butterfly-tag-plugins-plus@latest/lib/assets/carousel-touch.js"></script><!-- hexo injector head_end end --><meta name="generator" content="Hexo 7.2.0"></head><body><div id="loading-box"><div class="loading-left-bg"></div><div class="loading-right-bg"></div><div class="spinner-box"><div class="configure-border-1"><div class="configure-core"></div></div><div class="configure-border-2"><div class="configure-core"></div></div><div class="loading-word">加载中...</div></div></div><script>(()=>{
  const $loadingBox = document.getElementById('loading-box')
  const $body = document.body
  const preloader = {
    endLoading: () => {
      $body.style.overflow = ''
      $loadingBox.classList.add('loaded')
    },
    initLoading: () => {
      $body.style.overflow = 'hidden'
      $loadingBox.classList.remove('loaded')
    }
  }

  preloader.initLoading()
  window.addEventListener('load',() => { preloader.endLoading() })

  if (false) {
    document.addEventListener('pjax:send', () => { preloader.initLoading() })
    document.addEventListener('pjax:complete', () => { preloader.endLoading() })
  }
})()</script><div id="web_bg"></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="/img/head.jpg" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">4</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">5</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">3</div></a></div><hr class="custom-hr"/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 归档</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> 关于</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg fixed" id="page-header" style="background-image: url('/img/pagecover2.jpg')"><nav id="nav"><span id="blog-info"><a href="/" title="Eddie's Blog"></a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search" href="javascript:void(0);"><i class="fas fa-search fa-fw"></i><span> 搜索</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 归档</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> 关于</span></a></div></div><div id="toggle-menu"><a class="site-page" href="javascript:void(0);"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">RNA-seq分析流程</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2024-05-22T10:57:36.000Z" title="发表于 2024-05-22 18:57:36">2024-05-22</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2024-06-10T09:20:24.000Z" title="更新于 2024-06-10 17:20:24">2024-06-10</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E5%AD%A6%E4%B9%A0%F0%9F%93%9A/">学习📚</a></span></div><div class="meta-secondline"></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><p>记录一下学习RNA-seq分析的过程，整体流程参考，可能会补充一些其他的资料：</p>
<p><style><br>refer .a {<br>    color: white;<br>    font-size: 8px;<br>    text-decoration: none;<br>}</p>
<p>refer a:hover {<br>    color: blue;<br>}<br>&lt;/style&gt;</p>
<refer><ul>
<li> <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/518132262">RNA-seq入门实战（零）：RNA-seq流程前的准备——Linux与R的环境创建</a></li>
<li> <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/518134934">RNA-seq入门实战（一）：上游数据下载、格式转化和质控清洗</a></li>
<li> <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/518137593">RNA-seq入门实战（二）：上游数据的比对计数——Hisat2+ featureCounts 与 Salmon</a></li>
<li> <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/518140221">RNA-seq入门实战（三）：从featureCounts与Salmon输出文件获取counts矩阵</a></li>
<li> <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/518141288">RNA-seq入门实战（四）：差异分析前的准备——数据检查</a></li>
<li> <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/518142469">RNA-seq入门实战（五）：差异分析——DESeq2 edgeR limma的使用与比较</a></li>
<li> <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/518143539">RNA-seq入门实战（六）：GO、KEGG富集分析与enrichplot超全可视化攻略</a></li>
<li> <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/518144716">RNA-seq入门实战（七）：GSEA——基因集富集分析</a></li>
<li> <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/518145829">RNA-seq入门实战（八）：GSVA——基因集变异分析</a></li>
<li> <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/532037554">RNA-seq入门实战（九）：PPI蛋白互作网络构建（上）——STRING数据库的使用</a></li>
<li> <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/532038640">RNA-seq入门实战（十）：PPI蛋白互作网络构建（下）——Cytoscape软件的使用</a></li>
<li> <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/532039670">RNA-seq入门实战（十一）：WGCNA加权基因共表达网络分析——关联基因模块与表型</a></li>
</ul></refer>

<h1 id="背景知识"><a href="#背景知识" class="headerlink" title="背景知识"></a>背景知识</h1><hr>
<h1 id="上游数据下载、获取和质控分析"><a href="#上游数据下载、获取和质控分析" class="headerlink" title="上游数据下载、获取和质控分析"></a>上游数据下载、获取和质控分析</h1><h2 id="数据下载"><a href="#数据下载" class="headerlink" title="数据下载"></a>数据下载</h2><p>Source：NCBI-&gt;GEO数据库-&gt;SRA Run Select  <br><br>Packge: Linux Sra-tools prefetch方法    <br><br>Output: SRA数据<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">prefetch -O ./ $id &amp;</span><br></pre></td></tr></table></figure><br><details class="folding-tag" yellow><summary> 查看方法参数 </summary>
              <div class='content'>
              <ul><li>-O ： 下载目录</li><li>id : 下载数据SRA号，例：SRR12207280</li></ul>
              </div>
            </details></p>
<h2 id="SRA数据转fastq"><a href="#SRA数据转fastq" class="headerlink" title="SRA数据转fastq"></a>SRA数据转fastq</h2><p>Packge: Linux Sra-tools fasterq-dump/fastq-dump<br><br>    fasterq-dump速度更快，但不支持<code>--gzip</code><br><br>Input: *.sra    <br><br>Output: .fastq / .fastq.gz<br><br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fastq-dump --gzip --split-3 -O $&#123;outdirectory&#125; $&#123;SRRxxx&#125;.sra</span><br></pre></td></tr></table></figure><br><details class="folding-tag" yellow><summary> 查看方法参数 </summary>
              <div class='content'>
              <ul><li>—gzip: 输出gz格式压缩文件</li><li>—split-spot: 将双端测序分为两份,但是都放在同一个文件中</li><li>—split-files: 将双端测序分为两份,放在不同的文件,但是对于一方有而一方没有的reads直接丢弃</li><li>—split-3: 默认值，将双端测序分为两份,放在不同的文件,但是对于一方有而一方没有的reads会单独放在一个文件夹里</li></ul>
              </div>
            </details><br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fasterq-dump  -p -e 24 --split-3 -O $&#123;outdirectory&#125; $&#123;SRRxxx&#125;.sra</span><br></pre></td></tr></table></figure><br><details class="folding-tag" yellow><summary> 查看方法参数 </summary>
              <div class='content'>
              <ul><li>-P: 可以显示进程 </li><li>-e: 使用线程数目<br>note： —gzip不可用！！</li></ul>
              </div>
            </details><br>参考：<br><br><div class="reference-source"><span class="hidden-anchor" id="referfrom_[1] "></span><a class="reference-anchor" href="#referto_[1] ">[1] <div class="reference-anchor-up fa-solid fa-angles-up"></div></a><a class="reference-link" href=" https://zhuanlan.zhihu.com/p/591140275"> sra转fastq笔记</a></div></p>
<hr>
<h2 id="质量控制"><a href="#质量控制" class="headerlink" title="质量控制"></a>质量控制</h2><h3 id="原始数据质量查看"><a href="#原始数据质量查看" class="headerlink" title="原始数据质量查看"></a>原始数据质量查看</h3><h4 id="fastqc"><a href="#fastqc" class="headerlink" title="fastqc"></a>fastqc</h4><p>一款快速多线程地对测序数据进行质量评估（Quality Control）的质控软件，将生成html报告和报告压缩文件。<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] [-c contaminant file] seqfile1 .. seqfileN</span><br></pre></td></tr></table></figure><br><details class="folding-tag" yellow><summary> 查看方法参数 </summary>
              <div class='content'>
              <ul><li>-o: 输出目录</li><li>-noextract: 生成的报告默认会打包成1个压缩文件，使用这个参数是让程序不打包</li><li>-f: 强制输入文件格式，默认自动检测，有效输入bam,sam,bam_mapped,sam_mapped and fastq</li><li>-t: 线程数</li><li>-q: 安静运行模式</li></ul>
              </div>
            </details></p>
<h4 id="multiqc"><a href="#multiqc" class="headerlink" title="multiqc"></a>multiqc</h4><p>一种模块化工具，用于将多个样本的质量控制分析结果汇总到一份报告中，同样会生成html报告<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">multiqc [file_dir] --ignore [ignore_file]</span><br></pre></td></tr></table></figure><br><details class="folding-tag" yellow><summary> 查看方法参数 </summary>
              <div class='content'>
              <ul><li>[file_dir]:输入qc结果文件夹，将自动分析文件夹下的全部目录</li><li>[ignore_file]:</li></ul>
              </div>
            </details><br>原始数据质量判断戳这儿<br><div class="reference-source"><span class="hidden-anchor" id="referfrom_"></span><a class="reference-anchor" href="#referto_"><div class="reference-anchor-up fa-solid fa-angles-up"></div></a><a class="reference-link" href=" https://zhuanlan.zhihu.com/p/57628300"> fastqc报告查看</a></div><br><div class="reference-source"><span class="hidden-anchor" id="referfrom_"></span><a class="reference-anchor" href="#referto_"><div class="reference-anchor-up fa-solid fa-angles-up"></div></a><a class="reference-link" href=" https://www.jianshu.com/p/f83626fd1fa1"> multiqc报告查看</a></div></p>
<h3 id="质控清洗"><a href="#质控清洗" class="headerlink" title="质控清洗"></a>质控清洗</h3><h4 id="trim-galore"><a href="#trim-galore" class="headerlink" title="trim-galore"></a>trim-galore</h4><p>Trim Galore是对FastQC和Cutadapt的包装,适用于所有高通量测序，包括RRBS(Reduced Representation Bisulfite-Seq ), Illumina、Nextera和smallRNA测序平台的双端和单端数据，包括去除低质量碱基和3’ 末端的adapter以及长度控制等功能。<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">trim_galore -j 4  -q 25  --phred33 --length 35 \</span><br><span class="line">    --stringency 3  --paired --gzip -o [output_path] [file1] [file2]</span><br></pre></td></tr></table></figure><br><details class="folding-tag" yellow><summary> 查看方法参数 </summary>
              <div class='content'>
              <ul><li>-j: 使用线程数, 注意假设已使用Python 3并安装了Pigz，那么内核设置为4，实际使用内核是15，因此，最高设为4.</li><li>-1：设定Phred quality score阈值，默认为Phred 20 切除质量得分低于设定值的序列</li><li>—phred33/—phred64： 用于指定输入数据的质量得分编码方式，ASCII 33 是由 Illumina 1.8+ 和 Sanger 测序平台使用的，而 ASCII 64 是由 Illumina 1.3+ 和 Illumina 1.5+ 测序平台使用的，可以用FastQC检查类型。</li><li>—length: 设定输出reads长度阈值小于设定值会被抛弃，默认值20bp。</li><li>—stringency: </li><li>—paired：指定输入数据是配对端（paired-end）测序数据，如若是单端测序则不需此参数</li><li>—gzip：清洗后的数据zip打包</li><li>-o：输出路径</li><li>— trim-n：移除read一端的reads</li><li>—fastqc：使用默认选项对结果文件进行fastqc分析</li></ul>
              </div>
            </details></p>
<hr>
<h1 id="上游数据比对、计数"><a href="#上游数据比对、计数" class="headerlink" title="上游数据比对、计数"></a>上游数据比对、计数</h1><h2 id="hisat2-featurecount"><a href="#hisat2-featurecount" class="headerlink" title="hisat2+featurecount"></a>hisat2+featurecount</h2><h3 id="hisat2比对"><a href="#hisat2比对" class="headerlink" title="hisat2比对"></a>hisat2比对</h3><p>HISAT2是一种用于比对RNA序列数据的工具，在使用之前需要先在<a target="_blank" rel="noopener" href="https://daehwankimlab.github.io/hisat2/download/">获取索引文件</a>，或通过fasta文件自行建立。&lt;/br&gt;<br>索引文件的主要作用是加速比对过程。如果没有索引，HISAT2 需要对每一个读取和参考基因组的每一个位置进行比较，这将需要大量的时间。有了索引，HISAT2 可以快速地找到可能的比对位置，大大减少了比对所需的时间。此外，索引文件还可以包含其他的信息，如剪接位点（splice sites）和转录起始位点（transcription start sites）。这些信息可以帮助HISAT2更准确地比对RNA序列数据。<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">hisat2 -t -p 12 -x  [index] \</span><br><span class="line">-1 [file1]  \</span><br><span class="line">-2  [file2]  -S  [output.sam]</span><br></pre></td></tr></table></figure><br><details class="folding-tag" yellow><summary> 查看方法参数 </summary>
              <div class='content'>
              <ul><li>-x：参考基因组/转录组文件前缀数，若文件形如genome.1.ht2、genome.2.ht2，则只需要genome作为参数则可。</li><li>-1/-2：双端测序的第一二组文件，如果有多组数据，用逗号分隔</li><li>-U：单端测序数据</li><li>–sra-acc：SRA_ID，Hisat将自动下载并比对</li><li>-S：输出路径</li><li>-p：线程数</li></ul>
              </div>
            </details></p>
<h3 id="samtools处理"><a href="#samtools处理" class="headerlink" title="samtools处理"></a>samtools处理</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">samtools sort -@ 12 -o  id_sorted.bam  id.sam</span><br><span class="line">samtools index  -@  12  &#123;&#125;</span><br><span class="line">samtools flagstat -@ 12 id.bam &gt; id.flagstat</span><br></pre></td></tr></table></figure>
<details class="folding-tag" yellow><summary> 查看方法参数 </summary>
              <div class='content'>
              <ul><li>sort 排序</li><li>index 建立索引</li><li>flagstat 统计 BAM 文件中的比对结果<br>摆烂了 先不管这个的参数了</li></ul>
              </div>
            </details>
<h3 id="featureCounts-计数得到counts表达矩阵"><a href="#featureCounts-计数得到counts表达矩阵" class="headerlink" title="featureCounts 计数得到counts表达矩阵"></a>featureCounts 计数得到counts表达矩阵</h3><p>gtf索引获取网址(注意和之前的hisat2索引对应)：</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://www.gencodegenes.org/">GENCODE</a></li>
<li><a target="_blank" rel="noopener" href="https://hgdownload.soe.ucsc.edu/downloads.html">UCSC Genome</a></li>
</ul>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_"># </span><span class="language-bash">统计比对到每个genomic feature（外显子、基因等）中的<span class="built_in">read</span>数目</span></span><br><span class="line">featureCounts -T  12  -p  -a  xx.gtf.gz  -o  counts.txt  *.bam</span><br></pre></td></tr></table></figure>
<details class="folding-tag" yellow><summary> 查看方法参数 </summary>
              <div class='content'>
              <ul><li>-T: 线程数目</li><li>-p：只能用在paired-end的情况中，会统计fragment的特征而非read的</li><li>-a: 参考gtf文件名，支持Gzipped文件格式</li><li>-M: 统计多重map的read</li><li>-J：统计可变剪切</li><li>-G：当-J设置的时候，通过-G提供一个比对的时候使用的参考基因组文件，辅助寻找可变剪切</li></ul>
              </div>
            </details>
<div class="reference-source"><span class="hidden-anchor" id="referfrom_"></span><a class="reference-anchor" href="#referto_"><div class="reference-anchor-up fa-solid fa-angles-up"></div></a><a class="reference-link" href=" https://www.jianshu.com/p/9cc4e8657d62"> featurecounts的使用说明</a></div>
<h2 id="salmon-采用伪比对，直接对基因定量"><a href="#salmon-采用伪比对，直接对基因定量" class="headerlink" title="salmon-采用伪比对，直接对基因定量"></a>salmon-采用伪比对，直接对基因定量</h2><p>与hisat2不同，salmon不经过比对计数步骤而是直接对基因进行定量，如果不研究新转录本，用salmon方法可以更快更方便得到基因定量信息。按我自己的理解应该是hisat2是对全基因组进行比对，而salmon只对已知的基因注释比对。</p>
<h3 id="建立salmon索引"><a href="#建立salmon索引" class="headerlink" title="建立salmon索引"></a>建立salmon索引</h3><p>在<a target="_blank" rel="noopener" href="https://ftp.ensembl.org/pub/release-102/fasta/">ENSEMBL官网</a>获取cDNA参考文件<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">salmon index -t transcripts.fa -i transcripts_index -k 31</span><br></pre></td></tr></table></figure><br><details class="folding-tag" yellow><summary> 查看方法参数 </summary>
              <div class='content'>
              <ul><li>-t：转录本的fasta文件</li><li>-i：输出目录</li><li>-k：K-mers，默认值为31<h1 id="如果你的reads大于75bp，那么k设置为31是较好的选择，如果reads低于75可略微减少K值"><a href="#如果你的reads大于75bp，那么k设置为31是较好的选择，如果reads低于75可略微减少K值" class="headerlink" title="如果你的reads大于75bp，那么k设置为31是较好的选择，如果reads低于75可略微减少K值"></a>如果你的reads大于75bp，那么k设置为31是较好的选择，如果reads低于75可略微减少K值</h1></li></ul>
              </div>
            </details></p>
<h3 id="salmon定量"><a href="#salmon定量" class="headerlink" title="salmon定量"></a>salmon定量</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">#</span><span class="language-bash">双端测序数据reads表达量的估计</span></span><br><span class="line">salmon quant -i transcripts_index -l &lt;LIBTYPE&gt; -1 reads1.fq -2 reads2.fq -o transcripts_quant</span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash">单端测序数据reads表达量的估计</span></span><br><span class="line">salmon quant -i transcripts_index -l &lt;LIBTYPE&gt; -r reads.fq -o transcripts_quant</span><br></pre></td></tr></table></figure>
<details class="folding-tag" yellow><summary> 查看方法参数 </summary>
              <div class='content'>
              <ul><li>-i：index</li><li>-1/2：双端数据</li><li>-r：单端数据</li><li>-l：—libType，测序文库类型，用A指定automatically determine</li></ul>
              </div>
            </details>
<h3 id="获取基因ID转化的对应文件"><a href="#获取基因ID转化的对应文件" class="headerlink" title="获取基因ID转化的对应文件"></a>获取基因ID转化的对应文件</h3><p>原文-&gt;<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/9e62f9148932">Here</a><br>由于本次使用的为gencode或ensembl的gtf与cdna文件，因此最后得到的为ensembl_id (gene_id)和 transcript_id，形式为：ENSMUSG00000000001.1 ，而我们下游常用gene symbol进行展示，因此还需要从gtf注释文件中获取ensembl_id 、transcript_id与gene symbol的对应关系文件。<br><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">#</span><span class="language-bash">提取gtf注释文件中gene_id等与gene_name的对应关系,便于下游<span class="built_in">id</span>转换</span></span><br><span class="line">gtf=&quot;gencode.vM25.chr_patch_hapl_scaff.annotation.gtf&quot;</span><br><span class="line"><span class="meta prompt_"></span></span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash"><span class="comment">## gene_id to gene_name</span></span></span><br><span class="line">grep &#x27;gene_id&#x27; $gtf | awk -F &#x27;gene_id \&quot;&#x27; &#x27;&#123;print $2&#125;&#x27; |awk -F &#x27;\&quot;&#x27; &#x27;&#123;print $1&#125;&#x27; &gt;gene_id_tmp</span><br><span class="line">grep &#x27;gene_id&#x27; $gtf | awk -F &#x27;gene_name \&quot;&#x27; &#x27;&#123;print $2&#125;&#x27; |awk -F &#x27;\&quot;&#x27; &#x27;&#123;print $1&#125;&#x27; &gt;gene_name_tmp</span><br><span class="line">paste gene_id_tmp gene_name_tmp &gt;last_tmp</span><br><span class="line">uniq last_tmp &gt;g2s_vm25_gencode.txt</span><br><span class="line">rm *_tmp</span><br><span class="line"><span class="meta prompt_"></span></span><br><span class="line"><span class="meta prompt_">#</span><span class="language-bash"><span class="comment">## transcript_id to gene_name</span></span></span><br><span class="line">grep &#x27;transcript_id&#x27; $gtf | awk -F &#x27;transcript_id \&quot;&#x27; &#x27;&#123;print $2&#125;&#x27; |awk -F &#x27;\&quot;&#x27; &#x27;&#123;print $1&#125;&#x27; &gt;gene_id_tmp</span><br><span class="line">grep &#x27;transcript_id&#x27; $gtf | awk -F &#x27;gene_name \&quot;&#x27; &#x27;&#123;print $2&#125;&#x27; |awk -F &#x27;\&quot;&#x27; &#x27;&#123;print $1&#125;&#x27; &gt;gene_name_tmp</span><br><span class="line">paste gene_id_tmp gene_name_tmp &gt;last_tmp</span><br><span class="line">uniq last_tmp &gt;t2s_vm25_gencode.txt</span><br><span class="line">rm *_tmp</span><br></pre></td></tr></table></figure></p>
<h2 id="从featureCounts输出文件中获取counts矩阵"><a href="#从featureCounts输出文件中获取counts矩阵" class="headerlink" title="从featureCounts输出文件中获取counts矩阵"></a>从featureCounts输出文件中获取counts矩阵</h2><h3 id="从GEO上获取SraRunTable后，读取counts并读取分组信息"><a href="#从GEO上获取SraRunTable后，读取counts并读取分组信息" class="headerlink" title="从GEO上获取SraRunTable后，读取counts并读取分组信息"></a>从GEO上获取SraRunTable后，读取counts并读取分组信息</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>tidyverse<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>data.table<span class="punctuation">)</span></span><br><span class="line">counts_file <span class="operator">&lt;-</span> <span class="string">&quot;./data/count/counts.txt&quot;</span></span><br><span class="line">rawcounts <span class="operator">&lt;-</span> fread<span class="punctuation">(</span>counts_file<span class="punctuation">,</span>header <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">,</span>data.table <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span> <span class="comment">#header=T-&gt;第一列是列名</span></span><br><span class="line">colnames<span class="punctuation">(</span>rawcounts<span class="punctuation">)</span></span><br><span class="line">counts <span class="operator">&lt;-</span> rawcounts<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">7</span><span class="operator">:</span>ncol<span class="punctuation">(</span>rawcounts<span class="punctuation">)</span><span class="punctuation">]</span> <span class="comment">#截取样本基因表达量的counts部分作为counts </span></span><br><span class="line">rownames<span class="punctuation">(</span>counts<span class="punctuation">)</span> <span class="operator">&lt;-</span> rawcounts<span class="operator">$</span>Geneid</span><br><span class="line">colnames<span class="punctuation">(</span>counts<span class="punctuation">)</span></span><br><span class="line"><span class="comment">#gsub(&#x27;被替换的&#x27;,&#x27;替换&#x27;,data) 将data中的&#x27;被替换的&#x27;全部换成&#x27;替换&#x27;，如果为&#x27;&#x27;则等价于删去</span></span><br><span class="line">colnames<span class="punctuation">(</span>counts<span class="punctuation">)</span> <span class="operator">&lt;-</span> gsub<span class="punctuation">(</span><span class="string">&#x27;/XXX/RNA-seq/data/align/&#x27;</span><span class="punctuation">,</span><span class="string">&#x27;&#x27;</span><span class="punctuation">,</span><span class="comment">#删除样品名前缀</span></span><br><span class="line">                    gsub<span class="punctuation">(</span><span class="string">&#x27;_sorted.bam&#x27;</span><span class="punctuation">,</span><span class="string">&#x27;&#x27;</span><span class="punctuation">,</span>  colnames<span class="punctuation">(</span>counts<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment">#删除样品名后缀</span></span><br><span class="line">runtable <span class="operator">&lt;-</span> fread<span class="punctuation">(</span><span class="string">&quot;./data/SraRunTable&quot;</span><span class="punctuation">)</span> <span class="comment">#从GEO上下载的metadata，包括每个样本的元数据信息，例如样本的处理条件、生物学重复、疾病状态等。</span></span><br><span class="line">name_list <span class="operator">&lt;-</span> runtable<span class="operator">$</span>source_name<span class="punctuation">[</span>match<span class="punctuation">(</span>colnames<span class="punctuation">(</span>counts<span class="punctuation">)</span><span class="punctuation">,</span>runtable<span class="operator">$</span>Run<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment">#为分组编号</span></span><br><span class="line">name_list <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span>name_list<span class="punctuation">,</span> <span class="string">&quot;_&quot;</span><span class="punctuation">,</span> ave<span class="punctuation">(</span><span class="built_in">rep</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="built_in">length</span><span class="punctuation">(</span>name_list<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span> name_list<span class="punctuation">,</span> FUN <span class="operator">=</span> <span class="built_in">seq_along</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">group_list <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;naive&quot;</span><span class="punctuation">,</span><span class="string">&quot;naive&quot;</span><span class="punctuation">,</span><span class="string">&quot;primed&quot;</span><span class="punctuation">,</span><span class="string">&quot;primed&quot;</span><span class="punctuation">)</span></span><br><span class="line">batchinfo <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>row.names<span class="operator">=</span>colnames<span class="punctuation">(</span>counts<span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">                   name_list<span class="operator">=</span>name_list<span class="punctuation">,</span></span><br><span class="line">                   group_list<span class="operator">=</span>group_list<span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>counts<span class="punctuation">)</span> <span class="operator">&lt;-</span> name_list</span><br><span class="line">groupinfo <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span>row.names<span class="operator">=</span>colnames<span class="punctuation">(</span>counts<span class="punctuation">)</span><span class="punctuation">,</span> <span class="comment">#构建样品名与分组对应的数据框</span></span><br><span class="line">                 group_list<span class="operator">=</span>group_list<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<h3 id="将counts原始矩阵转换为TPM"><a href="#将counts原始矩阵转换为TPM" class="headerlink" title="将counts原始矩阵转换为TPM"></a>将counts原始矩阵转换为TPM</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#提取Geneid、Length(转录本长度)</span></span><br><span class="line">geneid_efflen <span class="operator">&lt;-</span> subset<span class="punctuation">(</span>rawcounts<span class="punctuation">,</span>select <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;Geneid&quot;</span><span class="punctuation">,</span><span class="string">&quot;Length&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>geneid_efflen<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;geneid&quot;</span><span class="punctuation">,</span><span class="string">&quot;efflen&quot;</span><span class="punctuation">)</span>  </span><br><span class="line"></span><br><span class="line"><span class="comment">### 取出counts中geneid对应的efflen</span></span><br><span class="line">efflen <span class="operator">&lt;-</span> geneid_efflen<span class="punctuation">[</span>match<span class="punctuation">(</span>rownames<span class="punctuation">(</span>counts<span class="punctuation">)</span><span class="punctuation">,</span>geneid_efflen<span class="operator">$</span>geneid<span class="punctuation">)</span><span class="punctuation">,</span><span class="string">&quot;efflen&quot;</span><span class="punctuation">]</span></span><br><span class="line">counts2TPM <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="punctuation">(</span>count<span class="operator">=</span>count<span class="punctuation">,</span> efflength<span class="operator">=</span>efflen<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  RPK <span class="operator">&lt;-</span> count<span class="operator">/</span><span class="punctuation">(</span>efflength<span class="operator">/</span><span class="number">1000</span><span class="punctuation">)</span>   <span class="comment">#每千碱基reads (Reads Per Kilobase) 长度标准化</span></span><br><span class="line">  PMSC_rpk <span class="operator">&lt;-</span> <span class="built_in">sum</span><span class="punctuation">(</span>RPK<span class="punctuation">)</span><span class="operator">/</span><span class="number">1e6</span>        <span class="comment">#RPK的每百万缩放因子 (“per million” scaling factor ) 深度标准化</span></span><br><span class="line">  RPK<span class="operator">/</span>PMSC_rpk              </span><br><span class="line"><span class="punctuation">&#125;</span>  </span><br><span class="line"></span><br><span class="line">tpm <span class="operator">&lt;-</span> as.data.frame<span class="punctuation">(</span>apply<span class="punctuation">(</span>counts<span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span>counts2TPM<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">colSums<span class="punctuation">(</span>tpm<span class="punctuation">)</span>   </span><br></pre></td></tr></table></figure>
<div class="note disabled blue flat"><p><div class="tip "><p>为什么要将Counts转变成TPM?</p>
</div><br>TPM (Transcripts Per Kilobase Million)  每千个碱基的转录每百万映射读取的Transcripts&lt;/br&gt;<br>首先,我们要明确计算TPM是在干嘛，其实就是在归一化。&lt;/br&gt;<br>那它归一对象是谁呢？&lt;/br&gt;</p>
<script type="math/tex; mode=display">
TPM = \frac{\frac{Count}{Length/1000}}{\sum{\frac{Count}{Length/1000}}} \times 10^6</script><p>从TPM的计算公式中我们不难发现涉及到了Count和Length：<span class='p red'>reads</span></p>
<ul>
<li>Count：Count 是测序的原始读取数。由于不同的样本可能有不同的测序深度（即测序的总读取数不同），我们需要对 Count 进行归一化，以消除测序深度的影响。这样，我们就可以在同一水平上比较<span class='p red'>不同样本</span>的基因表达量。</li>
<li>Length：在 RNA-seq 实验中，RNA 分子被打断成较短的片段进行测序。因此，长的基因会有更多的片段可以被测序，从而在测序数据中产生更多的 reads。但这并不意味着长的基因在生物体中就一定表达得更多。因此，我们需要对基因长度进行归一化，以消除基因长度的影响。这样，我们就可以在同一水平上比较<span class='p red'>不同长度</span>的基因的表达量。</li>
</ul>
</div>
<h3 id="基因ID转换并过滤低表达基因"><a href="#基因ID转换并过滤低表达基因" class="headerlink" title="基因ID转换并过滤低表达基因"></a>基因ID转换并过滤低表达基因</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">g2s <span class="operator">&lt;-</span> fread<span class="punctuation">(</span><span class="string">&quot;./data/g2s_vm25_gencode.txt&quot;</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span>data.table <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>g2s<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;geneid&quot;</span><span class="punctuation">,</span><span class="string">&quot;symbol&quot;</span><span class="punctuation">)</span></span><br><span class="line">symbol <span class="operator">&lt;-</span> g2s<span class="punctuation">[</span>match<span class="punctuation">(</span>rownames<span class="punctuation">(</span>counts<span class="punctuation">)</span><span class="punctuation">,</span>g2s<span class="operator">$</span>geneid<span class="punctuation">)</span><span class="punctuation">,</span><span class="string">&quot;symbol&quot;</span><span class="punctuation">]</span> <span class="comment">#匹配counts行名对应的symbol</span></span><br><span class="line">table<span class="punctuation">(</span>duplicated<span class="punctuation">(</span>symbol<span class="punctuation">)</span><span class="punctuation">)</span>  <span class="comment">#统计重复基因名</span></span><br><span class="line">counts <span class="operator">&lt;-</span> aggregate<span class="punctuation">(</span>counts<span class="punctuation">,</span> by<span class="operator">=</span><span class="built_in">list</span><span class="punctuation">(</span>symbol<span class="punctuation">)</span><span class="punctuation">,</span> FUN<span class="operator">=</span><span class="built_in">sum</span><span class="punctuation">)</span></span><br><span class="line">counts <span class="operator">&lt;-</span> column_to_rownames<span class="punctuation">(</span>counts<span class="punctuation">,</span><span class="string">&#x27;Group.1&#x27;</span><span class="punctuation">)</span></span><br><span class="line">tpm <span class="operator">&lt;-</span> aggregate<span class="punctuation">(</span>tpm<span class="punctuation">,</span> by<span class="operator">=</span><span class="built_in">list</span><span class="punctuation">(</span>symbol<span class="punctuation">)</span><span class="punctuation">,</span> FUN<span class="operator">=</span><span class="built_in">sum</span><span class="punctuation">)</span> <span class="comment">###使用aggregat 将symbol列中的相同基因进行合并 </span></span><br><span class="line">tpm <span class="operator">&lt;-</span> column_to_rownames<span class="punctuation">(</span>tpm<span class="punctuation">,</span><span class="string">&#x27;Group.1&#x27;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">### filter</span></span><br><span class="line">filter_gene <span class="operator">&lt;-</span> rowSums<span class="punctuation">(</span>counts<span class="operator">&gt;</span><span class="number">1</span><span class="punctuation">)</span> <span class="operator">&gt;=</span> <span class="number">2</span></span><br><span class="line">table<span class="punctuation">(</span>filter_gene<span class="punctuation">)</span> </span><br><span class="line">counts_filt <span class="operator">&lt;-</span> counts<span class="punctuation">[</span>filter_gene<span class="punctuation">,</span> <span class="punctuation">]</span> <span class="comment">#替换counts为筛选后的基因矩阵（保留较高表达量的基因）</span></span><br><span class="line">tpm_filt <span class="operator">&lt;-</span> tpm<span class="punctuation">[</span>filter_gene<span class="punctuation">,</span> <span class="punctuation">]</span></span><br></pre></td></tr></table></figure>
<h2 id="从salmon输出文件中获取counts矩阵"><a href="#从salmon输出文件中获取counts矩阵" class="headerlink" title="从salmon输出文件中获取counts矩阵"></a>从salmon输出文件中获取counts矩阵</h2><p>这部分开始不直接传大段代码了，感觉意义不大，反正也是从别人那照抄，也不方便自己日后查找，主要放一些核心代码、记录方法和思路&lt;/br&gt;<br>tximport从各种外部软件输出中导入转录水平估计值，并选择性地将丰度、计数和转录本长度汇总到基因水平（默认）或输出转录水平基质<br><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">tximport<span class="punctuation">(</span>files<span class="punctuation">,</span> type <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;none&quot;</span><span class="punctuation">,</span> <span class="string">&quot;kallisto&quot;</span><span class="punctuation">,</span> <span class="string">&quot;salmon&quot;</span><span class="punctuation">,</span> <span class="string">&quot;sailfish&quot;</span><span class="punctuation">,</span> <span class="string">&quot;rsem&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">\</span></span><br><span class="line">countsFromAbundance <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;no&quot;</span><span class="punctuation">,</span> <span class="string">&quot;scaledTPM&quot;</span><span class="punctuation">,</span> <span class="string">&quot;lengthScaledTPM&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="punctuation">\</span></span><br><span class="line">tx2gene <span class="operator">=</span> <span class="literal">NULL</span><span class="punctuation">,</span> reader <span class="operator">=</span> read.delim<span class="punctuation">)</span></span><br></pre></td></tr></table></figure><br><details class="folding-tag" yellow><summary> 查看方法参数 </summary>
              <div class='content'>
              <ul><li>files：salmon输出的quant.sf文件</li><li>type：用于生成丰度的软件类型，例如”kallisto”, “salmon”, “sailfish”, “rsem”</li><li>tx2gene:将转录本 ID（第 1 列）链接到基因 ID（第 2 列）的两列 Data.Frame（顺序不可换）</li><li>reader：可以用外部的read方法来提高提取速度</li></ul>
              </div>
            </details><br><div class="reference-source"><span class="hidden-anchor" id="referfrom_"></span><a class="reference-anchor" href="#referto_"><div class="reference-anchor-up fa-solid fa-angles-up"></div></a><a class="reference-link" href=" https://rdocumentation.org/packages/tximport/versions/1.0.3/topics/tximport"> tximport官方文档</a></div></p>
<hr>
<h1 id="差异表达分析"><a href="#差异表达分析" class="headerlink" title="差异表达分析"></a>差异表达分析</h1><h2 id="数据检查"><a href="#数据检查" class="headerlink" title="数据检查"></a>数据检查</h2><h3 id="归一化"><a href="#归一化" class="headerlink" title="归一化"></a>归一化</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#log对tpm归一化</span></span><br><span class="line">dat <span class="operator">&lt;-</span> <span class="built_in">log</span><span class="punctuation">(</span>tpm<span class="operator">+</span><span class="number">1</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">#构建DESeq对象</span></span><br><span class="line">dds <span class="operator">&lt;-</span> DESeqDataSetFromMatrix<span class="punctuation">(</span>countData <span class="operator">=</span> readcounts<span class="punctuation">,</span></span><br><span class="line">                              colData <span class="operator">=</span> groupinfo<span class="punctuation">,</span></span><br><span class="line">                              design <span class="operator">=</span> <span class="operator">~</span> group_list<span class="punctuation">)</span></span><br><span class="line">rld <span class="operator">&lt;-</span> rlog<span class="punctuation">(</span>dds<span class="punctuation">,</span> blind<span class="operator">=</span><span class="literal">FALSE</span><span class="punctuation">)</span> <span class="comment">#rlog进行归一化</span></span><br><span class="line">dat <span class="operator">&lt;-</span> as.data.frame<span class="punctuation">(</span>assay<span class="punctuation">(</span>rld<span class="punctuation">)</span><span class="punctuation">)</span> </span><br></pre></td></tr></table></figure>
<h3 id="作图检查"><a href="#作图检查" class="headerlink" title="作图检查"></a>作图检查</h3><ul>
<li>箱图<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">boxplot<span class="punctuation">(</span>dat<span class="punctuation">,</span>col<span class="operator">=</span>color<span class="punctuation">,</span> ylab<span class="operator">=</span><span class="string">&quot;dat&quot;</span><span class="punctuation">,</span> main<span class="operator">=</span><span class="string">&quot; normalized data &quot;</span><span class="punctuation">,</span></span><br><span class="line">        outline <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span> notch <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure></li>
<li>样本hclust图<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"></span><br></pre></td></tr></table></figure></li>
<li>距离热图<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">sampleDists <span class="operator">&lt;-</span> dist<span class="punctuation">(</span>t<span class="punctuation">(</span>dat<span class="punctuation">)</span><span class="punctuation">)</span>   <span class="comment">#dist默认计算矩阵行与行的距离， 因此需要转置</span></span><br><span class="line">sampleDistMatrix <span class="operator">&lt;-</span> as.matrix<span class="punctuation">(</span>sampleDists<span class="punctuation">)</span>  </span><br><span class="line">colors <span class="operator">&lt;-</span> colorRampPalette<span class="punctuation">(</span> rev<span class="punctuation">(</span>brewer.pal<span class="punctuation">(</span><span class="number">9</span><span class="punctuation">,</span> <span class="string">&quot;Blues&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">)</span><span class="punctuation">(</span><span class="number">255</span><span class="punctuation">)</span>  <span class="comment">#选取热图的颜色</span></span><br><span class="line">p0 <span class="operator">&lt;-</span> pheatmap<span class="operator">::</span>pheatmap<span class="punctuation">(</span>sampleDistMatrix<span class="punctuation">,</span></span><br><span class="line">                         fontsize<span class="operator">=</span><span class="number">7</span><span class="punctuation">,</span></span><br><span class="line">                         clustering_distance_rows<span class="operator">=</span>sampleDists<span class="punctuation">,</span></span><br><span class="line">                         clustering_distance_cols<span class="operator">=</span>sampleDists<span class="punctuation">,</span></span><br><span class="line">                         angle_col<span class="operator">=</span><span class="number">45</span><span class="punctuation">,</span></span><br><span class="line">                         col<span class="operator">=</span>colors<span class="punctuation">)</span></span><br></pre></td></tr></table></figure></li>
<li>PCA图<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">group_list <span class="operator">&lt;-</span> groupinfo<span class="operator">$</span>group_list</span><br><span class="line">dat.pca <span class="operator">&lt;-</span> PCA<span class="punctuation">(</span>t<span class="punctuation">(</span>dat<span class="punctuation">)</span> <span class="punctuation">,</span> graph <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span> </span><br><span class="line">percentVar <span class="operator">&lt;-</span> dat.pca<span class="operator">$</span>eig<span class="punctuation">[</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">]</span></span><br><span class="line">pca <span class="operator">&lt;-</span> fviz_pca_ind<span class="punctuation">(</span>dat.pca<span class="punctuation">,</span></span><br><span class="line">                    title <span class="operator">=</span> <span class="string">&quot;Principal Component Analysis&quot;</span><span class="punctuation">,</span></span><br><span class="line">                    legend.title <span class="operator">=</span> <span class="string">&quot;Groups&quot;</span><span class="punctuation">,</span></span><br><span class="line">                    geom.ind <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;point&quot;</span><span class="punctuation">,</span> <span class="string">&quot;text&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> </span><br><span class="line">                    pointsize <span class="operator">=</span> <span class="number">1.5</span><span class="punctuation">,</span></span><br><span class="line">                    labelsize <span class="operator">=</span> <span class="number">4</span><span class="punctuation">,</span></span><br><span class="line">                    col.ind <span class="operator">=</span> group_list<span class="punctuation">,</span> <span class="comment"># 分组上色</span></span><br><span class="line">                    axes.linetype<span class="operator">=</span><span class="literal">NA</span><span class="punctuation">,</span>  </span><br><span class="line">                    mean.point<span class="operator">=</span><span class="built_in">F</span> <span class="comment"># 去除分组中心点</span></span><br><span class="line"><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">  coord_fixed<span class="punctuation">(</span>ratio <span class="operator">=</span> <span class="number">1</span><span class="punctuation">)</span> <span class="operator">+</span>  <span class="comment"># 坐标轴的纵横比</span></span><br><span class="line">  xlab<span class="punctuation">(</span>paste0<span class="punctuation">(</span><span class="string">&quot;PC1 (&quot;</span><span class="punctuation">,</span> <span class="built_in">round</span><span class="punctuation">(</span>percentVar<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;%)&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">+</span>  </span><br><span class="line">  ylab<span class="punctuation">(</span>paste0<span class="punctuation">(</span><span class="string">&quot;PC2 (&quot;</span><span class="punctuation">,</span> <span class="built_in">round</span><span class="punctuation">(</span>percentVar<span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;%)&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span>  </span><br></pre></td></tr></table></figure></li>
<li>差异基因热图 <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"></span><br></pre></td></tr></table></figure></li>
<li>相关性热图<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"></span><br></pre></td></tr></table></figure></li>
</ul>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta"><i class="fas fa-circle-user fa-fw"></i>文章作者: </span><span class="post-copyright-info"><a href="https://EddieLjx.github.io">Eddie</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta"><i class="fas fa-square-arrow-up-right fa-fw"></i>文章链接: </span><span class="post-copyright-info"><a href="https://eddieljx.github.io/2024/05/22/RNA-seq%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/">https://eddieljx.github.io/2024/05/22/RNA-seq%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta"><i class="fas fa-circle-exclamation fa-fw"></i>版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="https://EddieLjx.github.io" target="_blank">Eddie's Blog</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%94%9F%E4%BF%A1/">生信</a><a class="post-meta__tags" href="/tags/%E4%BB%A3%E7%A0%81/">代码</a></div><div class="post_share"><div class="social-share" data-image="/img/pagecover2.jpg" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.3/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.3/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2024/10/18/%E8%AF%AD%E8%A8%80%E6%A8%A1%E5%9E%8B%E5%9C%A8%E7%94%9F%E7%89%A9%E7%A0%94%E7%A9%B6%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8/" title="语言模型在生物研究中的应用"><img class="cover" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://cdn.jsdelivr.net/gh/EddieLjx/BlogImg/LM_Review_NatureMethods/Cover_2.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">语言模型在生物研究中的应用</div></div></a></div><div class="next-post pull-right"><a href="/2024/05/11/share1/" title="还没开始复现 这只是一个测试"><img class="cover" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="/img/9.jpg" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">还没开始复现 这只是一个测试</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="/img/head.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">Leihjeunyihn</div><div class="author-info__description">不如现在就迈一步 在迷路的同时领悟</div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">4</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">5</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">3</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/EddieLjx"><i class="fab fa-github"></i><span>Follow Me</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="https://github.com/EddieLjx" target="_blank" title="Github"><i class="fab fa-github" style="color: #24292e;"></i></a><a class="social-icon" href="mailto:eddie_ljx@163.com" target="_blank" title="Email"><i class="fas fa-envelope" style="color: #4a7dbe;"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content">This is Eddie's Blog</div></div><div class="sticky_layout"><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2024/10/18/%E8%AF%AD%E8%A8%80%E6%A8%A1%E5%9E%8B%E5%9C%A8%E7%94%9F%E7%89%A9%E7%A0%94%E7%A9%B6%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8/" title="语言模型在生物研究中的应用"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://cdn.jsdelivr.net/gh/EddieLjx/BlogImg/LM_Review_NatureMethods/Cover_2.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="语言模型在生物研究中的应用"/></a><div class="content"><a class="title" href="/2024/10/18/%E8%AF%AD%E8%A8%80%E6%A8%A1%E5%9E%8B%E5%9C%A8%E7%94%9F%E7%89%A9%E7%A0%94%E7%A9%B6%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8/" title="语言模型在生物研究中的应用">语言模型在生物研究中的应用</a><time datetime="2024-10-18T11:48:00.000Z" title="发表于 2024-10-18 19:48:00">2024-10-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/05/22/RNA-seq%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/" title="RNA-seq分析流程"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="/img/pagecover2.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="RNA-seq分析流程"/></a><div class="content"><a class="title" href="/2024/05/22/RNA-seq%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/" title="RNA-seq分析流程">RNA-seq分析流程</a><time datetime="2024-05-22T10:57:36.000Z" title="发表于 2024-05-22 18:57:36">2024-05-22</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/05/11/share1/" title="还没开始复现 这只是一个测试"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="/img/9.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="还没开始复现 这只是一个测试"/></a><div class="content"><a class="title" href="/2024/05/11/share1/" title="还没开始复现 这只是一个测试">还没开始复现 这只是一个测试</a><time datetime="2024-05-11T08:30:00.000Z" title="发表于 2024-05-11 16:30:00">2024-05-11</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/05/10/welcome/" title="欢迎光临Eddie的blog"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="/img/pagecover1.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="欢迎光临Eddie的blog"/></a><div class="content"><a class="title" href="/2024/05/10/welcome/" title="欢迎光临Eddie的blog">欢迎光临Eddie的blog</a><time datetime="2024-05-10T02:30:00.000Z" title="发表于 2024-05-10 10:30:00">2024-05-10</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2024 By Leihjeunyihn</div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="translateLink" type="button" title="简繁转换">簡</button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button id="go-up" type="button" title="回到顶部"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js?v=4.13.0"></script><script src="/js/main.js?v=4.13.0"></script><script src="/js/tw_cn.js?v=4.13.0"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui@5.0.33/dist/fancybox/fancybox.umd.min.js"></script><script src="https://cdn.jsdelivr.net/npm/instant.page@5.2.0/instantpage.min.js" type="module"></script><script src="https://cdn.jsdelivr.net/npm/vanilla-lazyload@17.8.8/dist/lazyload.iife.min.js"></script><script src="https://cdn.jsdelivr.net/npm/node-snackbar@0.1.16/dist/snackbar.min.js"></script><div class="js-pjax"><script>if (!window.MathJax) {
  window.MathJax = {
    tex: {
      inlineMath: [['$', '$'], ['\\(', '\\)']],
      tags: 'ams'
    },
    chtml: {
      scale: 1.1
    },
    options: {
      renderActions: {
        findScript: [10, doc => {
          for (const node of document.querySelectorAll('script[type^="math/tex"]')) {
            const display = !!node.type.match(/; *mode=display/)
            const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display)
            const text = document.createTextNode('')
            node.parentNode.replaceChild(text, node)
            math.start = {node: text, delim: '', n: 0}
            math.end = {node: text, delim: '', n: 0}
            doc.math.push(math)
          }
        }, '']
      }
    }
  }
  
  const script = document.createElement('script')
  script.src = 'https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.min.js'
  script.id = 'MathJax-script'
  script.async = true
  document.head.appendChild(script)
} else {
  MathJax.startup.document.state(0)
  MathJax.texReset()
  MathJax.typesetPromise()
}</script><link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/katex.min.css"><script src="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/contrib/copy-tex.min.js"></script><script>(() => {
  document.querySelectorAll('#article-container span.katex-display').forEach(item => {
    btf.wrap(item, 'div', { class: 'katex-wrap'})
  })
})()</script></div><script id="click-show-text" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.3/dist/click-show-text.min.js" data-mobile="false" data-text="富强,民主,文明,和谐,自由,平等,公正,法治,爱国,敬业,诚信,友善" data-fontsize="20px" data-random="true" async="async"></script><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">搜索</span><span id="loading-status"></span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="is-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  数据库加载中</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="搜索文章" type="text"/></div></div><hr/><div id="local-search-results"></div><div id="local-search-stats-wrap"></div></div></div><div id="search-mask"></div><script src="/js/search/local-search.js?v=4.13.0"></script></div></div><!-- hexo injector body_end start --> <script data-pjax>if(document.getElementById('recent-posts') && (location.pathname ==='all'|| 'all' ==='all')){
    var parent = document.getElementById('recent-posts');
    var child = '<div class="recent-post-item" style="width:100%;height: auto"><div id="catalog_magnet"><div class="magnet_item"><a class="magnet_link" href="https://EddieLjx.github.io/tags/蛋白质设计/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">👓 Eddieの蛋白质设计学习笔记 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://EddieLjx.github.io/tags/机器学习/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">👩‍💻 Eddieの机器学习笔记 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://EddieLjx.github.io/tags/深度学习/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">💡 Eddieの深度学习笔记 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://EddieLjx.github.io/tags/生信/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">🧬 Eddieの生信笔记 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://EddieLjx.github.io/tags/代码/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">🖥️ Eddieの代码笔记 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item" style="visibility: hidden"></div><a class="magnet_link_more"  href="https://EddieLjx.github.io/tags" style="flex:1;text-align: center;margin-bottom: 10px;">查看更多...</a></div></div>';
    console.log('已挂载magnet')
    parent.insertAdjacentHTML("afterbegin",child)}
     </script><style>#catalog_magnet{flex-wrap: wrap;display: flex;width:100%;justify-content:space-between;padding: 10px 10px 0 10px;align-content: flex-start;}.magnet_item{flex-basis: calc(50% - 5px);background: #095477;margin-bottom: 10px;border-radius: 8px;transition: all 0.2s ease-in-out;}.magnet_item:hover{background: #166d95}.magnet_link_more{color:#555}.magnet_link{color:black}.magnet_link:hover{color:white}@media screen and (max-width: 600px) {.magnet_item {flex-basis: 100%;}}.magnet_link_context{display:flex;padding: 10px;font-size:16px;transition: all 0.2s ease-in-out;}.magnet_link_context:hover{padding: 10px 20px;}</style>
    <style></style><script data-pjax>
  function butterfly_footer_beautify_injector_config(){
    var parent_div_git = document.getElementById('footer-wrap');
    var item_html = '<div id="workboard"></div><p id="ghbdages"><a class="github-badge" target="_blank" href="https://hexo.io/" style="margin-inline:5px" data-title="博客框架为Hexo_v6.2.0" title=""><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://img.shields.io/badge/Frame-Hexo-blue?style=flat&amp;logo=hexo" alt=""/></a><a class="github-badge" target="_blank" href="https://butterfly.js.org/" style="margin-inline:5px" data-title="主题版本Butterfly_v4.3.1" title=""><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://img.shields.io/badge/Theme-Butterfly-6513df?style=flat&amp;logo=bitdefender" alt=""/></a><a class="github-badge" target="_blank" href="https://github.com/" style="margin-inline:5px" data-title="本站项目由Github托管" title=""><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://img.shields.io/badge/Source-Github-d021d6?style=flat&amp;logo=GitHub" alt=""/></a><a class="github-badge" target="_blank" href="http://creativecommons.org/licenses/by-nc-sa/4.0/" style="margin-inline:5px" data-title="本站采用知识共享署名-非商业性使用-相同方式共享4.0国际许可协议进行许可" title=""><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://img.shields.io/badge/Copyright-BY--NC--SA%204.0-d42328?style=flat&amp;logo=Claris" alt=""/></a></p>';
    console.log('已挂载butterfly_footer_beautify')
    parent_div_git.insertAdjacentHTML("beforeend",item_html)
    }
  var elist = 'null'.split(',');
  var cpage = location.pathname;
  var epage = 'all';
  var flag = 0;

  for (var i=0;i<elist.length;i++){
    if (cpage.includes(elist[i])){
      flag++;
    }
  }

  if ((epage ==='all')&&(flag == 0)){
    butterfly_footer_beautify_injector_config();
  }
  else if (epage === cpage){
    butterfly_footer_beautify_injector_config();
  }
  </script><script async src="/js/runtime.js"></script><script data-pjax>
    function butterfly_categories_card_injector_config(){
      var parent_div_git = document.getElementById('recent-posts');
      var item_html = '<style>li.categoryBar-list-item{width:32.3%;}.categoryBar-list{max-height: 190px;overflow:auto;}.categoryBar-list::-webkit-scrollbar{width:0!important}@media screen and (max-width: 650px){.categoryBar-list{max-height: 160px;}}</style><div class="recent-post-item" style="height:auto;width:100%;padding:0px;"><div id="categoryBar"><ul class="categoryBar-list"><li class="categoryBar-list-item" style="background:url(/img/15.jpg);"> <a class="categoryBar-list-link" href="categories/学习📚/">学习📚</a><span class="categoryBar-list-count">1</span><span class="categoryBar-list-descr"></span></li><li class="categoryBar-list-item" style="background:url(/img/8.jpg);"> <a class="categoryBar-list-link" href="categories/文章复现👨‍💻/">文章复现👨‍💻</a><span class="categoryBar-list-count">1</span><span class="categoryBar-list-descr"></span></li><li class="categoryBar-list-item" style="background:url(/img/16.jpg);"> <a class="categoryBar-list-link" href="categories/文献阅读📖/">文献阅读📖</a><span class="categoryBar-list-count">1</span><span class="categoryBar-list-descr"></span></li></ul></div></div>';
      console.log('已挂载butterfly_categories_card')
      parent_div_git.insertAdjacentHTML("afterbegin",item_html)
      }
    if( document.getElementById('recent-posts') && (location.pathname ==='all'|| 'all' ==='all')){
    butterfly_categories_card_injector_config()
    }
  </script><script async src="//at.alicdn.com/t/font_2032782_8d5kxvn09md.js"></script><!-- hexo injector body_end end --></body></html>