<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>语言模型在生物研究中的应用 | Eddie's Blog</title><meta name="author" content="Leihjeunyihn"><meta name="copyright" content="Leihjeunyihn"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="ffffff"><meta name="description" content="Information        Language models for biological research: a primer           Author:Elana Simon, Kyle Swanson &amp; James Zou           Jounal:Nature Methods     Year:2024       Methods and Backgrounds">
<meta property="og:type" content="article">
<meta property="og:title" content="语言模型在生物研究中的应用">
<meta property="og:url" content="https://eddieljx.github.io/2024/10/18/%E8%AF%AD%E8%A8%80%E6%A8%A1%E5%9E%8B%E5%9C%A8%E7%94%9F%E7%89%A9%E7%A0%94%E7%A9%B6%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8/index.html">
<meta property="og:site_name" content="Eddie&#39;s Blog">
<meta property="og:description" content="Information        Language models for biological research: a primer           Author:Elana Simon, Kyle Swanson &amp; James Zou           Jounal:Nature Methods     Year:2024       Methods and Backgrounds">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://cdn.jsdelivr.net/gh/EddieLjx/BlogImg/LM_Review_NatureMethods/Cover_2.png">
<meta property="article:published_time" content="2024-10-18T11:48:00.000Z">
<meta property="article:modified_time" content="2024-10-18T11:48:00.000Z">
<meta property="article:author" content="Leihjeunyihn">
<meta property="article:tag" content="深度学习">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://cdn.jsdelivr.net/gh/EddieLjx/BlogImg/LM_Review_NatureMethods/Cover_2.png"><link rel="shortcut icon" href="/img/favicon-16x16.png"><link rel="canonical" href="https://eddieljx.github.io/2024/10/18/%E8%AF%AD%E8%A8%80%E6%A8%A1%E5%9E%8B%E5%9C%A8%E7%94%9F%E7%89%A9%E7%A0%94%E7%A9%B6%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="stylesheet" href="/css/index.css?v=4.13.0"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.5.1/css/all.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/node-snackbar@0.1.16/dist/snackbar.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui@5.0.33/dist/fancybox/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: {"path":"/search.xml","preload":false,"top_n_per_article":1,"unescape":false,"languages":{"hits_empty":"找不到您查询的内容：${query}","hits_stats":"共找到 ${hits} 篇文章"}},
  translate: {"defaultEncoding":2,"translateDelay":0,"msgToTraditionalChinese":"繁","msgToSimplifiedChinese":"简"},
  noticeOutdate: undefined,
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":230},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: {"chs_to_cht":"你已切换为繁体中文","cht_to_chs":"你已切换为简体中文","day_to_night":"你已切换为深色模式","night_to_day":"你已切换为浅色模式","bgLight":"#49b1f5","bgDark":"#1f1f1f","position":"top-left"},
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid@4.11.1/dist/infinitegrid.min.js',
    buttonText: '加载更多'
  },
  isPhotoFigcaption: false,
  islazyload: true,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: true,
  },
  autoDarkmode: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '语言模型在生物研究中的应用',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2024-10-18 19:48:00'
}</script><script>(win=>{
      win.saveToLocal = {
        set: (key, value, ttl) => {
          if (ttl === 0) return
          const now = Date.now()
          const expiry = now + ttl * 86400000
          const item = {
            value,
            expiry
          }
          localStorage.setItem(key, JSON.stringify(item))
        },
      
        get: key => {
          const itemStr = localStorage.getItem(key)
      
          if (!itemStr) {
            return undefined
          }
          const item = JSON.parse(itemStr)
          const now = Date.now()
      
          if (now > item.expiry) {
            localStorage.removeItem(key)
            return undefined
          }
          return item.value
        }
      }
    
      win.getScript = (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        script.onerror = reject
        script.onload = script.onreadystatechange = function() {
          const loadState = this.readyState
          if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
          script.onload = script.onreadystatechange = null
          resolve()
        }

        Object.keys(attr).forEach(key => {
          script.setAttribute(key, attr[key])
        })

        document.head.appendChild(script)
      })
    
      win.getCSS = (url, id = false) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onerror = reject
        link.onload = link.onreadystatechange = function() {
          const loadState = this.readyState
          if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
          link.onload = link.onreadystatechange = null
          resolve()
        }
        document.head.appendChild(link)
      })
    
      win.activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', 'ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
        if (t === 'dark') activateDarkMode()
        else if (t === 'light') activateLightMode()
      
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
      const detectApple = () => {
        if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
          document.documentElement.classList.add('apple')
        }
      }
      detectApple()
    })(window)</script><link rel="stylesheet" href="/css/iconfont.css"><link rel="stylesheet" href="/css/custom.css" media="defer" onload="this.media='all'"><!-- hexo injector head_end start -->
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.12.0/dist/katex.min.css">

<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/hexo-math@4.0.0/dist/style.css">
<link rel="stylesheet" href="/css/runtime.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-categories-card@1.0.0/lib/categorybar.css"><link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-tag-plugins-plus@latest/lib/assets/font-awesome-animation.min.css" media="defer" onload="this.media='all'"><link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-tag-plugins-plus@latest/lib/tag_plugins.css" media="defer" onload="this.media='all'"><script src="https://npm.elemecdn.com/hexo-butterfly-tag-plugins-plus@latest/lib/assets/carousel-touch.js"></script><!-- hexo injector head_end end --><meta name="generator" content="Hexo 7.2.0"></head><body><div id="loading-box"><div class="loading-left-bg"></div><div class="loading-right-bg"></div><div class="spinner-box"><div class="configure-border-1"><div class="configure-core"></div></div><div class="configure-border-2"><div class="configure-core"></div></div><div class="loading-word">加载中...</div></div></div><script>(()=>{
  const $loadingBox = document.getElementById('loading-box')
  const $body = document.body
  const preloader = {
    endLoading: () => {
      $body.style.overflow = ''
      $loadingBox.classList.add('loaded')
    },
    initLoading: () => {
      $body.style.overflow = 'hidden'
      $loadingBox.classList.remove('loaded')
    }
  }

  preloader.initLoading()
  window.addEventListener('load',() => { preloader.endLoading() })

  if (false) {
    document.addEventListener('pjax:send', () => { preloader.initLoading() })
    document.addEventListener('pjax:complete', () => { preloader.endLoading() })
  }
})()</script><div id="web_bg"></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="/img/head.jpg" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">4</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">5</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">3</div></a></div><hr class="custom-hr"/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 归档</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> 关于</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg fixed" id="page-header" style="background-image: url('https://cdn.jsdelivr.net/gh/EddieLjx/BlogImg/LM_Review_NatureMethods/Cover_2.png')"><nav id="nav"><span id="blog-info"><a href="/" title="Eddie's Blog"></a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search" href="javascript:void(0);"><i class="fas fa-search fa-fw"></i><span> 搜索</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 归档</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> 关于</span></a></div></div><div id="toggle-menu"><a class="site-page" href="javascript:void(0);"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">语言模型在生物研究中的应用</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2024-10-18T11:48:00.000Z" title="发表于 2024-10-18 19:48:00">2024-10-18</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2024-10-18T11:48:00.000Z" title="更新于 2024-10-18 19:48:00">2024-10-18</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB%F0%9F%93%96/">文献阅读📖</a></span></div><div class="meta-secondline"></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1 id="Information"><a href="#Information" class="headerlink" title="Information"></a>Information</h1><table style="border: 2px solid #3fc1c1; border-collapse: collapse;">
  <tr style="border: 2px solid #3fc1c1;">
    <td colspan="3" style="text-align:center; border: 2px solid #3fc1c1;"><h2><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41592-024-02354-y">Language models for biological research: a primer</a></h2></td>
  </tr>
  <tr style="border: 2px solid #3fc1c1;">
    <td colspan="3" style="border: 2px solid #3fc1c1;"><h3>Author:</h3><b>Elana Simon, Kyle Swanson & James Zou</b></td>
  </tr>
  <tr style="border: 2px solid #3fc1c1;">
    <td colspan="2" style="border: 2px solid #3fc1c1;"><h3>Jounal:</h3><b>Nature Methods</b></td>
    <td style="border: 2px solid #3fc1c1;"><h3>Year:</h3><b>2024</b></td>
  </tr>
</table>

<hr>
<h1 id="Methods-and-Backgrounds"><a href="#Methods-and-Backgrounds" class="headerlink" title="Methods and Backgrounds"></a>Methods and Backgrounds</h1><ol>
<li>Language Model: 语言模型可以学习序列中的复杂模式，例如句子中的单词和蛋白质中的氨基酸。由于这些模型大多是在异构序列集合上训练的，因此它们可以学习灵活的模式，并且可以适应性解决各种特定问题。由于其适应性，因此通常作为支持广泛下游应用程序的基础模型。其训练过程往往包括预训练和微调两个部分，其中对更大的数据集进行预训练(无监督)为模型提供了对数据的基本理解，从而能够在微调期间更有效地学习新目标。</li>
</ol>
<hr>
<h1 id="Abstract"><a href="#Abstract" class="headerlink" title="Abstract"></a>Abstract</h1><p>本文旨在介绍、回顾语言模型(基于Transformer架构)在生物学研究中的作用，其中预训练模型通过三种常见方法被用于研究：1.直接预测；2.embedding分析；3.迁移学习。</p>
<p style="text-align:center;">
<img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://cdn.jsdelivr.net/gh/EddieLjx/BlogImg/LM_Review_NatureMethods/Fig1.png" style="zoom:50%;margin:0,auto;">
Fig. 1: Approaches to using language models for biological research.
</p>

<p>在本篇Primer中，作者介绍了如何使用这三种方法调整自然语言和生物语言模型来解决生物学中的一系列问题，以及使用语言模型进行生物学建模时的最佳实践，包括何时使用每种方法的建议以及给定当前模型限制需要考虑的注意事项。</p>
<hr>
<h1 id="Article-Details"><a href="#Article-Details" class="headerlink" title="Article Details"></a>Article Details</h1><h2 id="Natural-Language-Models"><a href="#Natural-Language-Models" class="headerlink" title="Natural Language Models"></a>Natural Language Models</h2><p>自然语言模型，例如ChatGPT，Claude之类的，在大量的数据集上训练，其中包括PubMed等生物来源。因此，它们可以微调为专业模型（例如，BioBERT或Med-PaLM2），为生物学研究者提供帮助。值得注意的是，因为ChatGPT、Claude这样的模型使用的训练数据集更广泛，因此其效果甚至可能优于Specialist的模型。</p>
<p>目前，自然语言模型可以大大加速生物学研究人员的效率，帮助阅读整理文献、总结文献中陌生概念，并可帮助完成生物信息学分析代码。</p>
<p style="text-align:center;">
<img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://cdn.jsdelivr.net/gh/EddieLjx/BlogImg/LM_Review_NatureMethods/Fig2.png" style="zoom:50%;margin:0,auto;">
Fig. 2: Example uses of natural language models for biological research.
</p>

<h2 id="Biology-Language-Models"><a href="#Biology-Language-Models" class="headerlink" title="Biology Language Models"></a>Biology Language Models</h2><h3 id="Protein-Language-Models"><a href="#Protein-Language-Models" class="headerlink" title="Protein Language Models"></a>Protein Language Models</h3><h4 id="Importance-of-Large-scale-pretraining"><a href="#Importance-of-Large-scale-pretraining" class="headerlink" title="Importance of Large-scale pretraining"></a>Importance of Large-scale pretraining</h4><p>在蛋白质序列的大型数据集上预先训练的蛋白质语言模型可以学习捕获蛋白质进化约束和关键特性的表示。随后在较小的标记数据集上对这些模型进行微调，可以准确预测下游任务，例如稳定性、相互作用，甚至具有指定结构的序列设计。</p>
<h4 id="Application"><a href="#Application" class="headerlink" title="Application"></a>Application</h4><h5 id="Direct-Prediction"><a href="#Direct-Prediction" class="headerlink" title="Direct Prediction:"></a>Direct Prediction:</h5><p>根据这些模型的原始训练目标，可以使用这些模型直接预测每个氨基酸出现在序列中给定位置的概率。由于训练数据涵盖了已知功能性蛋白质序列的所有谱系，因此这些模型有效地学习了蛋白质进化的模式。如果不对突变的影响进行实验测量，该模型就可以隐式地了解哪些突变会对蛋白质功能有害，因为根据在整个进化过程中实证观察到的突变，它们被认为不太可能发生。因此，这些预测可以开箱即用地用于估计蛋白质编码突变的影响。</p>
<p>可以通过在给定位置特异性掩盖野生型氨基酸并要求模型在给定序列的其余部分来估算掩蔽位置来获得突变的可能性。如果根据语言模型，突变的可能性低于野生型氨基酸，则表明该突变可能是有害的。评估突变致病性的实验研究已经根据模型似然, 验证了这些估计。</p>
<p>将蛋白质序列建模为语言的一个好处是这些序列<b>不需要事先比对或注释</b>; 其他方法需要进化上对齐的蛋白质序列来预测突变的影响。蛋白质序列似然的语言模型估计也可用于估计蛋白质序列是否可能形成功能结构，这使得蛋白质语言模型能够评估和设计新的序列</p>
<h5 id="Embedding-Analysis"><a href="#Embedding-Analysis" class="headerlink" title="Embedding Analysis"></a>Embedding Analysis</h5><p>除了输出之外，蛋白质语言模型还提供有用的蛋白质嵌入。具体来说，当蛋白质序列通过模型运行时，可以提取模型对蛋白质中每个氨基酸的内部表示（嵌入）。然后，每个氨基酸的嵌入可以单独使用或组合成单个蛋白质表示。例如，先前的工作发现<b>聚类蛋白质序列嵌入可以识别同源蛋白质</b>。然后，可以根据每种蛋白质中单个氨基酸嵌入之间的相似性将这些同源蛋白构建成<b>多序列比对</b>。</p>
<h5 id="Transfer-Learning"><a href="#Transfer-Learning" class="headerlink" title="Transfer Learning"></a>Transfer Learning</h5><p>由于微调蛋白质语言模型的成本很高，因此目前，更多的用法是将Embedding取出，输入给另一个更小的模型，以完成下游任务，例如：预测蛋白质稳定性、病毒抗原突变的免疫逃逸以及使用少量标记数据预测错义变体的致病性。或者，有更新、更高效的微调技术，使计算资源较少的研究人员能够完全微调大型蛋白质语言模型。</p>
<h4 id="Protein-Structure-Models"><a href="#Protein-Structure-Models" class="headerlink" title="Protein Structure Models"></a>Protein Structure Models</h4><p>虽然蛋白质结构预测模型，如 AlphaFold2和 ESMFold，不是本引物的重点，但值得一提的是，将结构信息与蛋白质序列一起包含在蛋白质序列中以训练模型，就像在蛋白质结构预测模型中所做的那样，可以改善各种下游任务的蛋白质表示。蛋白质结构预测模型与语言模型一样，已被证明可以通过直接预测、嵌入分析和迁移学习广泛适用于各种下游应用。</p>
<h3 id="Single-cell-Language-Models"><a href="#Single-cell-Language-Models" class="headerlink" title="Single-cell Language Models"></a>Single-cell Language Models</h3><h4 id="Importance-of-Large-scale-pretraining-1"><a href="#Importance-of-Large-scale-pretraining-1" class="headerlink" title="Importance of Large-scale pretraining"></a>Importance of Large-scale pretraining</h4><p>单细胞基因表达数据提供了对单个细胞的细胞状态和功能的见解，但它们的高维性使解释具有挑战性。最近开发了 AI 方法来帮助分析这些复杂数据。公开可用的单细胞基因表达数据的增长允许在不同细胞类型、组织、生物体和实验条件下的数百万个转录组上训练语言模型。这些模型可以泛化到新的数据集，并且可以针对各种下游任务进行微调，包括单元类型注释和批量校正。</p>
<h4 id="Example-Geneformer"><a href="#Example-Geneformer" class="headerlink" title="Example: Geneformer"></a>Example: Geneformer</h4><p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41586-023-06139-9">Geneformer</a>就是这样一种单细胞语言模型。与许多其他生物语言模型一样，它具有经过训练的 transformer 架构，可为许多下游应用提供基因和细胞的表示。Geneformer 将<b>每个细胞表示为细胞中表达量最大的 2,048 个基因的列表</b>，根据 RNA 表达水平排序。训练过程与前面描述的蛋白质语言模型类似，因为基因的子集被掩盖，模型被训练以预测缺失的基因。为了按照表达水平的顺序正确预测缺失的基因，该模型必须<b>了解各种基因的表达水平之间的相互作用，并隐式学习细胞类型特异性模式和背景</b>。Geneformer 在跨越 40 种组织类型的 3000 万个单细胞转录组上进行了训练，这有助于它学习不同的表达模式。尽管 Geneformer 关注每个基因的相对表达水平，但其他单细胞语言模型使用替代公式。例如，单细胞语言模型 <a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41592-024-02201-0">scGPT</a>对定量表达值进行了预训练，从而实现了略有不同的下游应用。scGPT 还可以包括实验元数据，例如模态、批次和扰动条件。</p>
<h4 id="Application-1"><a href="#Application-1" class="headerlink" title="Application"></a>Application</h4><h5 id="Direct-Prediction-1"><a href="#Direct-Prediction-1" class="headerlink" title="Direct Prediction:"></a>Direct Prediction:</h5><p>单细胞语言模型的直接输出支持各种创造性的计算机实验。该模型可以通过获取单个细胞中按表达排序的原始基因列表、修改基因的顺序并量化这如何改变输出来估计遗传扰动对细胞的影响。例如，Geneformer 通过人为地将POU5F1、SOX2、KLF4 和 MYC 添加到细胞基因排名的顶部来模拟成纤维细胞的重编程，从而在计算上将细胞转移到诱导多能干细胞状态。同样，单细胞语言模型可以通过人为地从细胞的排名列表中删除基因并检查对细胞嵌入的影响来预测细胞对基因去除的敏感性。</p>
<h5 id="Embedding-Analysis-1"><a href="#Embedding-Analysis-1" class="headerlink" title="Embedding Analysis"></a>Embedding Analysis</h5><p>单细胞语言模型包含每个基因的嵌入，这些嵌入可以组合（例如，平均）以为每个细胞创建一个表示。这些细胞嵌入可用于聚类、可视化和细胞类型标记。由于训练数据的多样性和数量，这些模型可以在保持生物变异性的同时隐式减少批次效应，使它们能够从包含许多实验批次的数据集中识别细微的细胞亚型。</p>
<h5 id="Transfer-Learning-1"><a href="#Transfer-Learning-1" class="headerlink" title="Transfer Learning"></a>Transfer Learning</h5><p>尽管这些嵌入中可能会出现有意义的簇，例如细胞类型，但也可以对模型进行微调以预测单个细胞的特性。例如，可以微调单细胞语言模型，以整合跨实验条件的数据并预测细胞类型标签和细胞状态。它们甚至可以支持基因的多模态表示。例如，可以对 scGPT 进行微调，以包括染色质可及性和蛋白质丰度以及基因表达水平，从而实现跨模态的数据集集成。</p>
<h2 id="Multimodal-Language-Models-for-Biology"><a href="#Multimodal-Language-Models-for-Biology" class="headerlink" title="Multimodal Language Models for Biology"></a>Multimodal Language Models for Biology</h2><p>多模态模型可以跨多种数据模态（例如文本和图像）进行推理，从而使这些模型能够解决本质上涉及多种数据类型的任务。例如，病理学语言-图像预训练 （PLIP）在 Twitter 数据上进行了训练，以将病理图像与其字幕相匹配，使用户能够获取给定图像的字幕或查找给定文本描述的图像。同样，Med-PaLM Multimodal被训练为基于生物医学图像回答问题，而 MolT5被训练为根据分子结构用自然语言描述分子，包括有关其潜在生物学功能的信息。给定具有多种模态的数据点的足够示例，研究人员也可以为其他类型的生物数据训练多模态模型。</p>
<p>通过将生物文本的固定语言模型嵌入与来自其他领域的数据相结合，自然语言模型也可以应用于多模态设置，而无需额外训练。GenePT 为单细胞数据提供了一个示例。GenePT 利用语言模型的隐含基因组知识来嵌入细胞。具体来说，GenePT 使用 ChatGPT 嵌入细胞，首先使用 ChatGPT 嵌入来自 NCBI 的基因的文本描述，然后通过平均基于文本的基因嵌入来创建单细胞嵌入，按单细胞表达加权。在某些应用程序中，这些从自然语言模型派生的嵌入与来自生物语言模型（如 Geneformer）的嵌入相匹配或优于生物语言模型。类似的想法可以应用于生物学的其他领域;固定语言模型嵌入可以与来自替代模态的数据或模型合并，而无需额外训练。</p>
<h2 id="Best-practices-when-using-language-models-for-biology"><a href="#Best-practices-when-using-language-models-for-biology" class="headerlink" title="Best practices when using language models for biology"></a>Best practices when using language models for biology</h2><p style="text-align:center;">
<img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://cdn.jsdelivr.net/gh/EddieLjx/BlogImg/LM_Review_NatureMethods/Fig3.png" style="zoom:50%;margin:0,auto;">
Fig. 3: Choosing the right approach for adapting a language model.
</p>

<p>第一个决定是确定研究问题的目标。如果目标是数据探索，则可以使用嵌入分析方法以及降维和聚类等技术来揭示数据中的结构。但是，如果目标是根据数据进行预测，那么直接预测和迁移学习方法往往更有用。如果问题根据模型的训练数据和目标与模型的内在能力相匹配，那么直接预测方法是合适的，可能会根据目标修改输入。如果项目目标明显偏离模型的能力，或者如果有更特定于感兴趣任务的数据，那么迁移学习可能很有用。当有足够的数据和计算资源可用时，最好的方法可能是微调部分或全部语言模型。但是，如果数据或计算资源有限，另一种方法是使用语言模型计算新数据点的嵌入，并使用这些嵌入作为输入来训练单独的、通常较小的模型。此外，某些模型仅作为 Web 界面或应用程序编程接口 （API） 提供，这可能会限制其用于直接预测。其他具有开源代码和经过训练的模型参数的模型可用于嵌入分析或迁移学习。<br><div class="note disabled blue flat"><p><div class="tip "><p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41592-024-02354-y/tables/2">Biological language model resources</a></p>
</div><br>一些模型还附带了 Jupyter 笔记本或 Google Colab 笔记本，演示了如何将预训练模型用于各种应用程序。该表罗列了它们中的部分。</p>
</div></p>
<p>尽管语言模型可能非常强大，但它们具有用户应注意的重要限制。首先，语言模型仍然无法完美地解决许多生物问题，即使是它们最初被训练要解决的问题。这可能是由于模型的限制，这些模型可能无法学习控制训练数据的所有模式，以及训练数据的限制。训练数据可能过时或有干扰，并且在某些类型的数据代表性不足的情况下，它们可能存在差距。例如，自然语言模型仅包含其训练数据中包含的生物知识，因此它们将不知道训练后发现的发现。蛋白质语言模型通常在标准氨基酸上进行训练，因此无法反映输入表示中任何翻译后修饰的重要性。单细胞表达数据可能很嘈杂，而资金优先级可能会使数据量偏向于特定的组织类型和疾病状态，这两者都是影响模型性能的因素。</p>
<p>此外，为特定生物应用量身定制的模型有时仍优于生物语言模型，尤其是当先验知识可以为模型设计提供信息时。例如，包含蛋白质结构信息的方法已被证明优于使用在蛋白质序列上训练的语言模型的方法。最后，评估适应其他任务的语言模型的性能也需要小心。语言模型是在大量可能不会公开共享的数据上进行训练的，因此确保语言模型的训练数据与下游任务的测试数据之间没有数据泄漏可能具有挑战性。</p>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta"><i class="fas fa-circle-user fa-fw"></i>文章作者: </span><span class="post-copyright-info"><a href="https://EddieLjx.github.io">Eddie</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta"><i class="fas fa-square-arrow-up-right fa-fw"></i>文章链接: </span><span class="post-copyright-info"><a href="https://eddieljx.github.io/2024/10/18/%E8%AF%AD%E8%A8%80%E6%A8%A1%E5%9E%8B%E5%9C%A8%E7%94%9F%E7%89%A9%E7%A0%94%E7%A9%B6%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8/">https://eddieljx.github.io/2024/10/18/%E8%AF%AD%E8%A8%80%E6%A8%A1%E5%9E%8B%E5%9C%A8%E7%94%9F%E7%89%A9%E7%A0%94%E7%A9%B6%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta"><i class="fas fa-circle-exclamation fa-fw"></i>版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="https://EddieLjx.github.io" target="_blank">Eddie's Blog</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/">深度学习</a></div><div class="post_share"><div class="social-share" data-image="https://cdn.jsdelivr.net/gh/EddieLjx/BlogImg/LM_Review_NatureMethods/Cover_2.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.3/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.3/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="next-post pull-full"><a href="/2024/05/22/RNA-seq%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/" title="RNA-seq分析流程"><img class="cover" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="/img/pagecover2.jpg" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">RNA-seq分析流程</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="/img/head.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">Leihjeunyihn</div><div class="author-info__description">不如现在就迈一步 在迷路的同时领悟</div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">4</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">5</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">3</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/EddieLjx"><i class="fab fa-github"></i><span>Follow Me</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="https://github.com/EddieLjx" target="_blank" title="Github"><i class="fab fa-github" style="color: #24292e;"></i></a><a class="social-icon" href="mailto:eddie_ljx@163.com" target="_blank" title="Email"><i class="fas fa-envelope" style="color: #4a7dbe;"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content">This is Eddie's Blog</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#Information"><span class="toc-number">1.</span> <span class="toc-text">Information</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link"><span class="toc-number">1.1.</span> <span class="toc-text">Language models for biological research: a primer</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link"><span class="toc-number">1.1.1.</span> <span class="toc-text">Author:</span></a></li><li class="toc-item toc-level-3"><a class="toc-link"><span class="toc-number">1.1.2.</span> <span class="toc-text">Jounal:</span></a></li><li class="toc-item toc-level-3"><a class="toc-link"><span class="toc-number">1.1.3.</span> <span class="toc-text">Year:</span></a></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Methods-and-Backgrounds"><span class="toc-number">2.</span> <span class="toc-text">Methods and Backgrounds</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Abstract"><span class="toc-number">3.</span> <span class="toc-text">Abstract</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Article-Details"><span class="toc-number">4.</span> <span class="toc-text">Article Details</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#Natural-Language-Models"><span class="toc-number">4.1.</span> <span class="toc-text">Natural Language Models</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Biology-Language-Models"><span class="toc-number">4.2.</span> <span class="toc-text">Biology Language Models</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#Protein-Language-Models"><span class="toc-number">4.2.1.</span> <span class="toc-text">Protein Language Models</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#Importance-of-Large-scale-pretraining"><span class="toc-number">4.2.1.1.</span> <span class="toc-text">Importance of Large-scale pretraining</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Application"><span class="toc-number">4.2.1.2.</span> <span class="toc-text">Application</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#Direct-Prediction"><span class="toc-number">4.2.1.2.1.</span> <span class="toc-text">Direct Prediction:</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#Embedding-Analysis"><span class="toc-number">4.2.1.2.2.</span> <span class="toc-text">Embedding Analysis</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#Transfer-Learning"><span class="toc-number">4.2.1.2.3.</span> <span class="toc-text">Transfer Learning</span></a></li></ol></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Protein-Structure-Models"><span class="toc-number">4.2.1.3.</span> <span class="toc-text">Protein Structure Models</span></a></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Single-cell-Language-Models"><span class="toc-number">4.2.2.</span> <span class="toc-text">Single-cell Language Models</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#Importance-of-Large-scale-pretraining-1"><span class="toc-number">4.2.2.1.</span> <span class="toc-text">Importance of Large-scale pretraining</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Example-Geneformer"><span class="toc-number">4.2.2.2.</span> <span class="toc-text">Example: Geneformer</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Application-1"><span class="toc-number">4.2.2.3.</span> <span class="toc-text">Application</span></a><ol class="toc-child"><li class="toc-item toc-level-5"><a class="toc-link" href="#Direct-Prediction-1"><span class="toc-number">4.2.2.3.1.</span> <span class="toc-text">Direct Prediction:</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#Embedding-Analysis-1"><span class="toc-number">4.2.2.3.2.</span> <span class="toc-text">Embedding Analysis</span></a></li><li class="toc-item toc-level-5"><a class="toc-link" href="#Transfer-Learning-1"><span class="toc-number">4.2.2.3.3.</span> <span class="toc-text">Transfer Learning</span></a></li></ol></li></ol></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Multimodal-Language-Models-for-Biology"><span class="toc-number">4.3.</span> <span class="toc-text">Multimodal Language Models for Biology</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Best-practices-when-using-language-models-for-biology"><span class="toc-number">4.4.</span> <span class="toc-text">Best practices when using language models for biology</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2024/10/18/%E8%AF%AD%E8%A8%80%E6%A8%A1%E5%9E%8B%E5%9C%A8%E7%94%9F%E7%89%A9%E7%A0%94%E7%A9%B6%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8/" title="语言模型在生物研究中的应用"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://cdn.jsdelivr.net/gh/EddieLjx/BlogImg/LM_Review_NatureMethods/Cover_2.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="语言模型在生物研究中的应用"/></a><div class="content"><a class="title" href="/2024/10/18/%E8%AF%AD%E8%A8%80%E6%A8%A1%E5%9E%8B%E5%9C%A8%E7%94%9F%E7%89%A9%E7%A0%94%E7%A9%B6%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8/" title="语言模型在生物研究中的应用">语言模型在生物研究中的应用</a><time datetime="2024-10-18T11:48:00.000Z" title="发表于 2024-10-18 19:48:00">2024-10-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/05/22/RNA-seq%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/" title="RNA-seq分析流程"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="/img/pagecover2.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="RNA-seq分析流程"/></a><div class="content"><a class="title" href="/2024/05/22/RNA-seq%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B/" title="RNA-seq分析流程">RNA-seq分析流程</a><time datetime="2024-05-22T10:57:36.000Z" title="发表于 2024-05-22 18:57:36">2024-05-22</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/05/11/share1/" title="还没开始复现 这只是一个测试"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="/img/9.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="还没开始复现 这只是一个测试"/></a><div class="content"><a class="title" href="/2024/05/11/share1/" title="还没开始复现 这只是一个测试">还没开始复现 这只是一个测试</a><time datetime="2024-05-11T08:30:00.000Z" title="发表于 2024-05-11 16:30:00">2024-05-11</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/05/10/welcome/" title="欢迎光临Eddie的blog"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="/img/pagecover1.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="欢迎光临Eddie的blog"/></a><div class="content"><a class="title" href="/2024/05/10/welcome/" title="欢迎光临Eddie的blog">欢迎光临Eddie的blog</a><time datetime="2024-05-10T02:30:00.000Z" title="发表于 2024-05-10 10:30:00">2024-05-10</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2024 By Leihjeunyihn</div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="translateLink" type="button" title="简繁转换">簡</button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js?v=4.13.0"></script><script src="/js/main.js?v=4.13.0"></script><script src="/js/tw_cn.js?v=4.13.0"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui@5.0.33/dist/fancybox/fancybox.umd.min.js"></script><script src="https://cdn.jsdelivr.net/npm/instant.page@5.2.0/instantpage.min.js" type="module"></script><script src="https://cdn.jsdelivr.net/npm/vanilla-lazyload@17.8.8/dist/lazyload.iife.min.js"></script><script src="https://cdn.jsdelivr.net/npm/node-snackbar@0.1.16/dist/snackbar.min.js"></script><div class="js-pjax"><script>if (!window.MathJax) {
  window.MathJax = {
    tex: {
      inlineMath: [['$', '$'], ['\\(', '\\)']],
      tags: 'ams'
    },
    chtml: {
      scale: 1.1
    },
    options: {
      renderActions: {
        findScript: [10, doc => {
          for (const node of document.querySelectorAll('script[type^="math/tex"]')) {
            const display = !!node.type.match(/; *mode=display/)
            const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display)
            const text = document.createTextNode('')
            node.parentNode.replaceChild(text, node)
            math.start = {node: text, delim: '', n: 0}
            math.end = {node: text, delim: '', n: 0}
            doc.math.push(math)
          }
        }, '']
      }
    }
  }
  
  const script = document.createElement('script')
  script.src = 'https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.min.js'
  script.id = 'MathJax-script'
  script.async = true
  document.head.appendChild(script)
} else {
  MathJax.startup.document.state(0)
  MathJax.texReset()
  MathJax.typesetPromise()
}</script><link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/katex.min.css"><script src="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/contrib/copy-tex.min.js"></script><script>(() => {
  document.querySelectorAll('#article-container span.katex-display').forEach(item => {
    btf.wrap(item, 'div', { class: 'katex-wrap'})
  })
})()</script></div><script id="click-show-text" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.3/dist/click-show-text.min.js" data-mobile="false" data-text="富强,民主,文明,和谐,自由,平等,公正,法治,爱国,敬业,诚信,友善" data-fontsize="20px" data-random="true" async="async"></script><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">搜索</span><span id="loading-status"></span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="is-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  数据库加载中</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="搜索文章" type="text"/></div></div><hr/><div id="local-search-results"></div><div id="local-search-stats-wrap"></div></div></div><div id="search-mask"></div><script src="/js/search/local-search.js?v=4.13.0"></script></div></div><!-- hexo injector body_end start --> <script data-pjax>if(document.getElementById('recent-posts') && (location.pathname ==='all'|| 'all' ==='all')){
    var parent = document.getElementById('recent-posts');
    var child = '<div class="recent-post-item" style="width:100%;height: auto"><div id="catalog_magnet"><div class="magnet_item"><a class="magnet_link" href="https://EddieLjx.github.io/tags/蛋白质设计/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">👓 Eddieの蛋白质设计学习笔记 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://EddieLjx.github.io/tags/机器学习/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">👩‍💻 Eddieの机器学习笔记 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://EddieLjx.github.io/tags/深度学习/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">💡 Eddieの深度学习笔记 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://EddieLjx.github.io/tags/生信/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">🧬 Eddieの生信笔记 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://EddieLjx.github.io/tags/代码/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">🖥️ Eddieの代码笔记 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item" style="visibility: hidden"></div><a class="magnet_link_more"  href="https://EddieLjx.github.io/tags" style="flex:1;text-align: center;margin-bottom: 10px;">查看更多...</a></div></div>';
    console.log('已挂载magnet')
    parent.insertAdjacentHTML("afterbegin",child)}
     </script><style>#catalog_magnet{flex-wrap: wrap;display: flex;width:100%;justify-content:space-between;padding: 10px 10px 0 10px;align-content: flex-start;}.magnet_item{flex-basis: calc(50% - 5px);background: #095477;margin-bottom: 10px;border-radius: 8px;transition: all 0.2s ease-in-out;}.magnet_item:hover{background: #166d95}.magnet_link_more{color:#555}.magnet_link{color:black}.magnet_link:hover{color:white}@media screen and (max-width: 600px) {.magnet_item {flex-basis: 100%;}}.magnet_link_context{display:flex;padding: 10px;font-size:16px;transition: all 0.2s ease-in-out;}.magnet_link_context:hover{padding: 10px 20px;}</style>
    <style></style><script data-pjax>
  function butterfly_footer_beautify_injector_config(){
    var parent_div_git = document.getElementById('footer-wrap');
    var item_html = '<div id="workboard"></div><p id="ghbdages"><a class="github-badge" target="_blank" href="https://hexo.io/" style="margin-inline:5px" data-title="博客框架为Hexo_v6.2.0" title=""><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://img.shields.io/badge/Frame-Hexo-blue?style=flat&amp;logo=hexo" alt=""/></a><a class="github-badge" target="_blank" href="https://butterfly.js.org/" style="margin-inline:5px" data-title="主题版本Butterfly_v4.3.1" title=""><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://img.shields.io/badge/Theme-Butterfly-6513df?style=flat&amp;logo=bitdefender" alt=""/></a><a class="github-badge" target="_blank" href="https://github.com/" style="margin-inline:5px" data-title="本站项目由Github托管" title=""><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://img.shields.io/badge/Source-Github-d021d6?style=flat&amp;logo=GitHub" alt=""/></a><a class="github-badge" target="_blank" href="http://creativecommons.org/licenses/by-nc-sa/4.0/" style="margin-inline:5px" data-title="本站采用知识共享署名-非商业性使用-相同方式共享4.0国际许可协议进行许可" title=""><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" data-lazy-src="https://img.shields.io/badge/Copyright-BY--NC--SA%204.0-d42328?style=flat&amp;logo=Claris" alt=""/></a></p>';
    console.log('已挂载butterfly_footer_beautify')
    parent_div_git.insertAdjacentHTML("beforeend",item_html)
    }
  var elist = 'null'.split(',');
  var cpage = location.pathname;
  var epage = 'all';
  var flag = 0;

  for (var i=0;i<elist.length;i++){
    if (cpage.includes(elist[i])){
      flag++;
    }
  }

  if ((epage ==='all')&&(flag == 0)){
    butterfly_footer_beautify_injector_config();
  }
  else if (epage === cpage){
    butterfly_footer_beautify_injector_config();
  }
  </script><script async src="/js/runtime.js"></script><script data-pjax>
    function butterfly_categories_card_injector_config(){
      var parent_div_git = document.getElementById('recent-posts');
      var item_html = '<style>li.categoryBar-list-item{width:32.3%;}.categoryBar-list{max-height: 190px;overflow:auto;}.categoryBar-list::-webkit-scrollbar{width:0!important}@media screen and (max-width: 650px){.categoryBar-list{max-height: 160px;}}</style><div class="recent-post-item" style="height:auto;width:100%;padding:0px;"><div id="categoryBar"><ul class="categoryBar-list"><li class="categoryBar-list-item" style="background:url(/img/15.jpg);"> <a class="categoryBar-list-link" href="categories/学习📚/">学习📚</a><span class="categoryBar-list-count">1</span><span class="categoryBar-list-descr"></span></li><li class="categoryBar-list-item" style="background:url(/img/8.jpg);"> <a class="categoryBar-list-link" href="categories/文章复现👨‍💻/">文章复现👨‍💻</a><span class="categoryBar-list-count">1</span><span class="categoryBar-list-descr"></span></li><li class="categoryBar-list-item" style="background:url(/img/16.jpg);"> <a class="categoryBar-list-link" href="categories/文献阅读📖/">文献阅读📖</a><span class="categoryBar-list-count">1</span><span class="categoryBar-list-descr"></span></li></ul></div></div>';
      console.log('已挂载butterfly_categories_card')
      parent_div_git.insertAdjacentHTML("afterbegin",item_html)
      }
    if( document.getElementById('recent-posts') && (location.pathname ==='all'|| 'all' ==='all')){
    butterfly_categories_card_injector_config()
    }
  </script><script async src="//at.alicdn.com/t/font_2032782_8d5kxvn09md.js"></script><!-- hexo injector body_end end --></body></html>